### DISCUSSION TITLE
Support new TypeScript `satisfies` operator.

### DISCUSSION THREAD

**Discussion by alii** [vercel/next.js#40895] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.1.0: Mon Aug 29 20:11:49 PDT 2022; root:xnu-8792.40.74.0.1~20/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.9.0
      npm: 8.19.1
      Yarn: 3.2.0
      pnpm: N/A
    Relevant packages:
      next: 12.3.2-canary.7
      eslint-config-next: 12.3.1
      react: 18.2.0
      react-dom: 18.2.0

### Describe the Bug

When using the new `satisfies` TypeScript operator, the compiler won't build the app as it doesn't understand the syntax.

Documentation for operator: https://devblogs.microsoft.com/typescript/announcing-typescript-4-9-beta/#the-satisfies-operator

Code:
<img width="377" alt="CleanShot 2022-09-24 at 11 15 31@2x" src="https://user-images.githubusercontent.com/25351731/192092619-7e14af58-7917-4c4e-8a9d-73ff0a9d6cb7.png">

Result:
![CleanShot 2022-09-24 at 11 14 43@2x](https://user-images.githubusercontent.com/25351731/192092571-9c445314-98b2-46bd-a837-3694923aa2e4.png)

### Expected Behavior

The build works

### Link to reproduction

n/a, you just need to use the satisfies keyword

### To Reproduce

1) Write the following code in a Next.js app (or any code that uses `satisfies`).
```ts
export const posts = [
	new ServerlessOAuthPart2(),
	new OpenSource(),
	new Mochip(),
	new ZeroKbBlog(),
	new ServerlessDiscordOAuth(),
	new Goals(),
	new StrictTSConfig(),
] as const satisfies readonly Post[];
```
2) Run `yarn build`
3) cry because your build doesn't work and the end of the world nears

**Comment by SukkaW** [comment]
SWC compiler already supports TS 4.9: https://github.com/swc-project/swc/pull/5938

Waiting for Next.js to bump the version.

**Comment by vilkinsons** [comment]
Is there an ETA on this given TS 4.9's GA? It's been a couple of weeks.

**Comment by frdwhite24** [comment]
Any update on an ETA for this? Finally found the perfect use-case for the new `satisfies` keyword but am blocked by this 😢

**Comment by PabloSzx** [comment]
Would it be possible to update Next.js 12 to support TypeScript 4.9? We still can't update to Next.js 13 and wont be able to for quite some time

@ijjk

**Comment by vilkinsons** [comment]
Example of us [trying](https://github.com/hashintel/hash/blob/1ed6fd2a83baab0e3302b3c588bc9b76936c0c71/packages/hash/frontend/src/pages/%5Baccount-slug%5D/entities/%5Bentity-uuid%5D.page/entity-editor/links-section/link-table/use-create-get-cell-content.ts) and [failing](https://vercel.com/hashintel/hash/HsZRVmKVDAmAUrYdxcYT9dQvFhE3) to get this to work. Here’s the specific bit of code:
```
          case "linkTitle":
            return {
              kind: GridCellKind.Custom,
              readonly: true,
              allowOverlay: false,
              copyData: linkRow.linkTitle,
              data: {
                kind: "link-cell",
                linkRow,
              } satisfies LinkCellProps,
            };
```

The TS playground suggests `satisfies` in a nested object is valid, so this should work.

**Comment by algomachine007** [comment]
Good day, i have the same issue on NEXT 12, would be glad to find a solution
<img width="354" alt="Screenshot 2023-01-19 at 5 01 07 PM" src="https://user-images.githubusercontent.com/95980716/213675874-94367996-df69-45aa-a628-9befffa16336.png">

**Comment by danceroutine** [comment]
Plus one unfortunately. 12.2.5

![image](https://user-images.githubusercontent.com/9252682/221100362-7441d5e5-4006-420b-8007-56f263009d6a.png)

**Comment by CodeVachon** [comment]
Still an Issue on `12.3.4`

<img width="409" alt="Screenshot 2023-04-27 at 10 16 19" src="https://user-images.githubusercontent.com/5806093/234890150-37ac78e5-de4d-4fce-81a1-c4d6a9f1c473.png">

**Comment by lavelle** [comment]
Is there any way to pin a new version of SWC without waiting for next itself to update? I'm guessing not since they have their own SWC package. We are also on 12.3.4 and having this issue. I'm wondering if things like this will start to come to v13 only

**Comment by santialbo** [comment]
Why was this moved to a discussion? There's nothing to discuss here. Next.js is lagging behind in Typescript support.

**Comment by codinsonn** [comment]
Tried updating and pinning all my Babel versions to either the latest or `^7.22.5` versions through the "resolutions" field in `package.json`, but the problem persists. Checking which versions are installed with `yarn why {babel module}` states there's only one newly updated version for each of the ones I updated, so the resolutions themselves definitely work... just not to resolve the problem 😅

As I've seen that updating the babel versions does work in other, non-next.js projects, I'm inclined to assume that either:
- next.js seems to simply ignore the resolved babel modules (e.g. because it prebuilds it's own compiler somehow?)
- there's more steps to this than just applying the right typescript and babel versions

**Comment by piersolenski** [comment]
Any updates @balazsorban44?

**Comment by f0rmat1k** [comment]
You have `satisfies` in documentation https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-paths, but it still doesn't work...

**Comment by CalebJamesStevens** [comment]
12.3.4 still an issue

**Comment by lance13c** [comment]
`satisfies` is working in our repo using these configs. Other variables may be needed to get this working, but I believe these are the most impactful. Let me know if any of these configs work for you.

_next.config.js_

```js
...
experimental: {
  swcPlugins: [['next-superjson-plugin', {}]],
}
...
```

_package.json_
```json
...
"next": "13.4.12",
"@swc/core": "^1.3.59",
"next-superjson-plugin": "^0.5.10"
"typescript": "^5.1.6"
...
```

**Comment by luixo** [comment]
If it still doesn't work for you with babel - check that you don't have `@babel/core` lower than 7.20 installed somewhere down your dependency tree (or maybe in your own `package.json`).

**Comment by BitOfUniverse** [comment]
In case anyone still having issues and looking for a solution, try to upgrade next.js to the latest minor/patch version of v13:
`yarn upgrade next@v13.5.5` 
it should be enough, if it's not try also upgrade @babel/core
`yarn upgrade @babel/core@7.23.9`

**Comment by eliezra236** [comment]
Still having this issue in 12.3.4
Is there any fix / workaround?

**Comment by Alejandro-J-Caputto** [comment]
is the only possible solution to upgrade to Nextjs 13?

**Reply by Alejandro-J-Caputto** [reply]
Ok guys, my collage was able to figure out a workaround.
 If you are still running next.js < 13. Try to force the resolution for ` @next/swc-darwin-arm64`: 
`package.json`
``` 
  "resolutions": {
    {
    "@next/swc-darwin-arm64": "13.0.7",
    "@next/swc-darwin-x64": "13.0.7",
    "@next/swc-linux-x64-gnu": "13.0.7",
    "@next/swc-linux-x64-musl": "13.0.7"
  }
  },

**Reply by tran-simon** [reply]
didnt work for us

**Reply by RodrigoTomeES** [reply]
> didnt work for us

You MUST remove the lock file, node_modules and reinstall the dependencies
